OkvisSLAMSystem.cpp

```diff
@@ -151,29 +154,50 @@ namespace ark {
                     keyframe->descriptors_[cam_idx]=frame_data.data->descriptors[cam_idx];
                 }
 
-                for (int i = 0; i < sparse_map_vector.size()-1; i++) {
-                    const auto sparseMap = sparse_map_vector[i];
-                    if (sparseMap->detectLoopClosure(keyframe)) {
-                        active_map_index = i;
-                        // delete all the new map after active map
-                        sparse_map_vector.resize(active_map_index+1);
-                        cout << "Deleting newer maps after: " << active_map_index << endl;
-                        for (MapSparseMapDeletionHandler::const_iterator callback_iter = mMapSparseMapDeletionHandler.begin();
-                            callback_iter != mMapSparseMapDeletionHandler.end(); ++callback_iter) {
-                            const MapSparseMapDeletionHandler::value_type& pair = *callback_iter;
-                            pair.second(i);
+                // apply correction
+                keyframe->T_WS_ = correction_ * keyframe->T_WS_;
+                MapKeyFrame::Ptr loop_kf = nullptr;
+                Eigen::Affine3d transformEstimate;
+                if (detectLoopClosure(keyframe, loop_kf, transformEstimate)) {
+                    //cout << "Loop closure detected" << endl;
+                    addKeyFrameResult = true;
+                    for (int i = 0; i < sparse_map_vector.size()-1; i++) {
+                        if (sparse_map_vector[i] == nullptr)
+                            continue;
+                        auto sparseMap = sparse_map_vector[i];
+                        if (sparseMap->getKeyframe(loop_kf->frameId_) != nullptr) {
+                            //cout << "MapMerge: " << i << " with " << active_map_index << " and " << keyframe->frameId_ << " with " << loop_kf->frameId_ << endl;
+                            sparse_map_vector[active_map_index] = mergeMaps(
+                                    sparseMap, getActiveMap(), keyframe, loop_kf, transformEstimate);
+                            sparse_map_vector.erase(sparse_map_vector.begin() + i);
+                            active_map_index = sparse_map_vector.size() - 1;
+                            mapsMerged = true;
+                            break;
                         }
-                        break;
                     }
                 }
-
-                addKeyFrameResult = getActiveMap()->addKeyframe(keyframe);
-
+                
+                if (!mapsMerged) {
+                    getActiveMap()->addKeyframe(keyframe, loop_kf, transformEstimate);
```

Comment:
Is loop_kf ever not nullptr here? Can you just pass in nullptr to clarify this if so